C++ API parity: MaxUnpool1d

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27060 C++ API parity: LogSigmoid
* #27059 C++ API parity: LeakyReLU
* #27038 C++ API parity: Hardtanh
* #27035 C++ API parity: Hardshrink
* #27028 C++ API parity: ELU
* #27027 C++ API parity: MaxUnpool3d
* #26915 C++ API parity: MaxUnpool2d
* **#26896 C++ API parity: MaxUnpool1d**

Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

pbelevich committed:
C++ API parity: MaxUnpool1d

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"




[ghstack-poisoned]

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -550,6 +550,28 @@ TEST_F(ModulesTest, AdaptiveAvgPool3d) {
+  ASSERT_TRUE(torch::allclose(y,
+    torch::tensor({{{0, 2, 0, 4, 5}}}, torch::kFloat)));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 1, 5}));
+}
```
It would great to also test the C++ equivalent of the example in https://pytorch.org/docs/stable/nn.html#torch.nn.MaxUnpool1d, to show that MaxPool -> MaxUnpool works well.

yf225 replied:
Do you think we can add the test for MaxPool -> MaxUnpool? Maybe something like:
```cpp
  {
    MaxPool1d pool(MaxPool1dOptions(2).stride(2));
    MaxUnpool1d unpool(MaxUnpool1dOptions(2).stride(2));
    auto input = torch::tensor({{{1., 2., 3., 4., 5., 6., 7., 8.}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8.}}})));
  }
  {
    MaxPool1d pool(MaxPool1dOptions(2).stride(2));
    MaxUnpool1d unpool(MaxUnpool1dOptions(2).stride(2));
    auto input = torch::tensor({{{1., 2., 3., 4., 5., 6., 7., 8., 9.}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices, /*output_size=*/input.sizes()),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8., 0.}}})));
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8.}}})));
  }
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -149,6 +149,52 @@ inline Tensor adaptive_avg_pool3d(const Tensor& input,
+
+inline std::vector<int64_t> _unpool_output_size(const Tensor& input,
+  const IntArrayRef kernel_size, const IntArrayRef stride,
+  const IntArrayRef padding, const IntArrayRef output_size) {
```
I think we should use `const IntArrayRef&` instead of `const IntArrayRef` for all arguments here.

yf225 replied:
I think we should change `output_size` to `const c10::optional<IntArrayRef>&` to better match the Python version, because `output_size` being empty may not necessarily mean it's "None" (since a tensor with size `{}` is still a valid tensor, and it might be what the user wants).

yf225 commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -149,6 +149,52 @@ inline Tensor adaptive_avg_pool3d(const Tensor& input,
+}
+
+inline Tensor max_unpool1d(const Tensor& input, const Tensor& indices,
+  const IntArrayRef output_size, const MaxUnpool1dOptions& options) {
```
Similarly here, we should change `output_size` to `const c10::optional<IntArrayRef>&`.

yf225 replied:
I'd also suggest putting `output_size` last in the argument list with default value `c10::nullopt`, so that when the user calls this function, they don't need to specify the `output_size` value if they don't want to.

yf225 commented on torch/csrc/api/include/torch/nn/modules/pooling.h:
```diff
@@ -324,5 +324,42 @@ class TORCH_API AdaptiveAvgPool3dImpl :
+ public:
+  using MaxUnpoolImpl<1, MaxUnpool1dImpl>::MaxUnpoolImpl;
+  Tensor forward(const Tensor& input, const Tensor& indices,
+                 IntArrayRef output_size = {});
```
Similarly here, we should change `output_size` to `const c10::optional<IntArrayRef>&`.

yf225 reviewed (commented):
Thanks @pbelevich ! I left some comments.

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -149,6 +149,53 @@ inline Tensor adaptive_avg_pool3d(const Tensor& input,
+  } else {
+    std::vector<int64_t> output_size_;
+    if (output_size->size() == kernel_size.size() + 2) {
+      output_size_ = {output_size->begin() + 2, output_size->end()};
```
To better match the Python version, maybe we should write
```suggestion
      output_size_ = output_size->slice(2);
```

pbelevich replied:
output_size_ = output_size->slice(2).vec();

yf225 replied:
`output_size_ = output_size->slice(2).vec();` looks great. Do you plan to change it?

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks @pbelevich !

pbelevich committed:
Update on "C++ API parity: MaxUnpool1d"


Differential Revision: [D17627825](https://our.internmc.facebook.com/intern/diff/D17627825)

[ghstack-poisoned]
